minifi-nar-bundles/minifi-provenance-reporting-bundle/minifi-provenance-reporting-task/src/main/java/org/apache/nifi/minifi/provenance/reporting/ProvenanceReportingTask.java

```diff
@@ -0,0 +1,443 @@
+
+        siteToSiteClient = new SiteToSiteClient.Builder()
+            .url(destinationUrl)
+            .portName(context.getProperty(PORT_NAME).getValue())
+            .useCompression(context.getProperty(COMPRESS).asBoolean())
+            .eventReporter(eventReporter)
+            .sslContext(sslContext)
+            .timeout(context.getProperty(TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS), TimeUnit.MILLISECONDS)
+            .build();
+    }
+
+    @OnStopped
+    public void shutdown() throws IOException {
+        final SiteToSiteClient client = getClient();
+        if (client != null) {
+            client.close();
+        }
+    }
+
+    // this getter is intended explicitly for testing purposes
+    protected SiteToSiteClient getClient() {
+        return this.siteToSiteClient;
+    }
+
+    private String getComponentName(final ProcessGroupStatus status, final ProvenanceEventRecord event) {
+        if (status == null) {
+            return null;
+        }
+
+        final String componentId = event.getComponentId();
+        if (status.getId().equals(componentId)) {
+            return status.getName();
+        }
+
+        for (final ProcessorStatus procStatus : status.getProcessorStatus()) {
+            if (procStatus.getId().equals(componentId)) {
+                return procStatus.getName();
+            }
+        }
+
+        for (final PortStatus portStatus : status.getInputPortStatus()) {
+            if (portStatus.getId().equals(componentId)) {
+                return portStatus.getName();
+            }
+        }
+
+        for (final PortStatus portStatus : status.getOutputPortStatus()) {
+            if (portStatus.getId().equals(componentId)) {
+                return portStatus.getName();
+            }
+        }
+
+        for (final RemoteProcessGroupStatus rpgStatus : status.getRemoteProcessGroupStatus()) {
+            if (rpgStatus.getId().equals(componentId)) {
+                return rpgStatus.getName();
+            }
+        }
+
+        for (final ProcessGroupStatus childGroup : status.getProcessGroupStatus()) {
+            final String componentName = getComponentName(childGroup, event);
+            if (componentName != null) {
+                return componentName;
+            }
+        }
+
+        return null;
+    }
+
+    @Override
+    public void onTrigger(final ReportingContext context) {
+        final ProcessGroupStatus procGroupStatus = context.getEventAccess().getControllerStatus();
+        final String rootGroupName = procGroupStatus == null ? null : procGroupStatus.getName();
+
+        if (firstEventId < 0) {
+            Map<String, String> state;
+            try {
+                state = context.getStateManager().getState(Scope.LOCAL).toMap();
+            } catch (IOException e) {
+                getLogger().error("Failed to get state at start up due to {}", e);
+                return;
+            }
+            if (state.containsKey(LAST_EVENT_ID_KEY)) {
```

Comment:
Have to take into consideration how we provide a consistent view of provenance data in terms of provenance implementations such as the Volatile item which resets to 0 on a restart